src/mod_cache_users.erl

```diff
@@ -0,0 +1,236 @@
+%% That's fine, worst case this record will live a segment less than expected.
+put_member(ParentTab, Key, SegmentKey) ->
+    EtsSegment = ets:lookup_element(ParentTab, SegmentKey, 2),
+    ets:insert(EtsSegment, {Key}),
+    ok.
+
+del_member(_, _, '$end_of_table') ->
+    ok;
+del_member(ParentTab, Key, SegmentKey) ->
+    case ets:lookup(ParentTab, SegmentKey) of
+        [{_, EtsSegment}] ->
+            ets:delete(EtsSegment, Key),
+            del_member(ParentTab, Key, ets:next(ParentTab, SegmentKey));
+        [] ->
+            del_member(ParentTab, Key, ets:next(ParentTab, SegmentKey))
+    end.
+
+del_domain(_, _, '$end_of_table') ->
+    ok;
+del_domain(ParentTab, Domain, SegmentKey) ->
+    case ets:lookup(ParentTab, SegmentKey) of
+        [{_, EtsSegment}] ->
+            ets:match_delete(EtsSegment, {{Domain, '_'}}),
+            del_domain(ParentTab, Domain, ets:next(ParentTab, SegmentKey));
+        [] ->
+            del_domain(ParentTab, Domain, ets:next(ParentTab, SegmentKey))
+    end.
+
+-spec start_cache(mongooseim:host_type(), gen_mod:module_opts()) -> any().
+start_cache(HostType, Opts) ->
+    ParentTab = tbl_name(HostType),
+    Spec = #{id => ParentTab, start => {?MODULE, start_link, [ParentTab, Opts]},
+             restart => permanent, shutdown => 5000,
+             type => worker, modules => [?MODULE]},
+    {ok, _} = ejabberd_sup:start_child(Spec).
+
+-spec stop_cache(mongooseim:host_type()) -> any().
+stop_cache(HostType) ->
+    ok = ejabberd_sup:stop_child(tbl_name(HostType)).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+-record(cache_state, {tab :: atom(), ttl = timer:seconds(30) :: non_neg_integer()}).
+
+rotate_and_purge_last_segment(#cache_state{tab = ParentTab}) ->
+    First = ets:first(ParentTab),
+    Last = ets:last(ParentTab),
+    EtsSegment = case ets:info(ParentTab, size) of
+                     1 ->
+                         ets:lookup_element(ParentTab, First, 2);
+                     _ ->
+                         [{_, E}] = ets:take(ParentTab, First),
+                         true = ets:insert_new(ParentTab, {Last + 1, E}),
+                         E
+                 end,
+    ets:delete_all_objects(EtsSegment),
+    ok.
+
+-spec start_link(atom(), gen_mod:module_opts()) -> {ok, pid()}.
+start_link(ParentTab, Opts) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [ParentTab, Opts], []).
+
+init([ParentTab, Opts]) ->
+    TTL = gen_mod:get_opt(ttl, Opts, infinity),
+    N   = gen_mod:get_opt(number_of_segments, Opts, 1),
+    EtsOpts = [ordered_set, named_table, protected, {read_concurrency, true}],
+    ParentTab = ets:new(ParentTab, EtsOpts),
+    lists:foreach(
+      fun(I) ->
+              SegmentOpts = [set, public, {read_concurrency, true}, {write_concurrency, true}],
+              Ref = ets:new(ParentTab, SegmentOpts),
```

Comment:
that line is confusing. Could you use a DIFFERENT names for the parent (named) table and segment tables? 
Even if there is no logic in your code using that, the debugging of unnamed becomes kinda a pain. i.e. simple `ets:info(users_cache_my_cool_host_1).` would not work. Of course, there could be a work around, but it's not simple.

I would rather name the resources, that could potentially eat some memory ;)